include/wil/network.h

```diff
@@ -0,0 +1,1338 @@
+    inline ::wil::unique_addrinfo resolve_localhost_addresses()
+    {
+        constexpr auto* localhost_address_name_string = L"localhost";
+        ADDRINFOW* addrResult{};
+        if (::GetAddrInfoW(localhost_address_name_string, nullptr, nullptr, &addrResult) != 0)
+        {
+            THROW_WIN32(::WSAGetLastError());
+        }
+
+        return ::wil::unique_addrinfo{addrResult};
+    }
+#endif
+
+    struct WINSOCK_EXTENSION_FUNCTION_TABLE
+    {
+        LPFN_ACCEPTEX AcceptEx{nullptr};
+        LPFN_CONNECTEX ConnectEx{nullptr};
+        LPFN_DISCONNECTEX DisconnectEx{nullptr};
+        LPFN_GETACCEPTEXSOCKADDRS GetAcceptExSockaddrs{nullptr};
+        LPFN_TRANSMITFILE TransmitFile{nullptr};
+        LPFN_TRANSMITPACKETS TransmitPackets{nullptr};
+        LPFN_WSARECVMSG WSARecvMsg{nullptr};
+        LPFN_WSASENDMSG WSASendMsg{nullptr};
+    };
+
+    template <typename F>
+    struct socket_extension_function_table_t;
+    using winsock_extension_function_table = socket_extension_function_table_t<WINSOCK_EXTENSION_FUNCTION_TABLE>;
+    using rio_extension_function_table = socket_extension_function_table_t<RIO_EXTENSION_FUNCTION_TABLE>;
+
+    template <typename F>
+    struct socket_extension_function_table_t
+    {
+        static socket_extension_function_table_t load() WI_NOEXCEPT;
+
+        ~socket_extension_function_table_t() WI_NOEXCEPT = default;
+
+        // can copy, but the new object needs its own WSA reference count
+        // (getting a WSA reference count should be no-fail once the first reference it taken)
+        //
+        // IF the target could not get a WSA reference count
+        //   OR
+        // IF we couldn't get our own WSA reference count
+        //   (this should never happen the caller has a reference, but we failed to get a WSA reference)
+        // THEN
+        //   this object cannot carry forward any function pointers - it must show successfully loaded == false
+        socket_extension_function_table_t(const socket_extension_function_table_t& rhs) WI_NOEXCEPT
+            : wsa_reference_count{WSAStartup_nothrow()}
+        {
+            if (!wsa_reference_count || !rhs.wsa_reference_count)
+            {
+                ::memset(&f, 0, sizeof(f));
+            }
+            else
+            {
+                ::memcpy_s(&f, sizeof(f), &rhs.f, sizeof(rhs.f));
+            }
+        }
+
+        socket_extension_function_table_t& operator=(const socket_extension_function_table_t& rhs) WI_NOEXCEPT
+        {
+            if (!wsa_reference_count || !rhs.wsa_reference_count)
+            {
+                ::memset(&f, 0, sizeof(f));
+            }
+            else
+            {
+                ::memcpy_s(&f, sizeof(f), &rhs.f, sizeof(rhs.f));
+            }
+
+            return *this;
+        }
+
+        // Returns true if all functions were loaded, holding a WSAStartup reference
+        WI_NODISCARD explicit operator bool() const WI_NOEXCEPT;
+
+        F f{};

```

Comment:
Should probably give this a more descriptive name if it is to be public

Reply:
yeah. I'm trying to keep this name extremely short, since callers will invoke it like test_table.ft.AcceptEx(...).

And I didn't want to replicate every function in the function table to a method on that object. So they have to just reference that table when they want to call a function.

Reply:
Something to possibly also consider doing to reduce the typing overhead (and therefore reducing the burden of the name here) is providing an `operator->` that returns a pointer to the table. E.g. your snippet becomes `test_table->AcceptEx(...)`. That's starting to dip your toes into "feature abuse" territory, but IMO it's not _that bad_

Another thing to consider is to derive this type from the table type, so all data members are accessible directly off the object. E.g. your snippet now becomes `test_table.AcceptEx(...)`. Your `memset` and `memcpy` calls become a _little_ bit more complex, but `static_cast` is your friend.

Reply:
Great ideas. Thanks!